ADR-183: Favorite places (#188)
decentraland/adr: content/ADR-183-favorite-places.md

In a future iteration it will be possible to favorite places from the minimap HUD and also from the map parcel info pop-up, currently not possible due to technical limitations of the renderer map handling and Places API restriction to curated places.
More in the limitations section.

## Solution

### Data storing

After an initial discussion about whether to store favorites in the user profile or in Places, we are more incline to store everything in Places, more informations about the decision journey at the end.

### Favorite retrieval

The Places API will provide the required informations through:

- An endpoint to obtain the (paginated if we decide for a big maximum amount) list of favorites.
- An endpoint to add a new favorite for the current user (already available)
- An endpoint to remove a favorite for the current user (already available)
In additon a change will be required in the current places data structure in order to show the favorite information in the explore places list with a boolean. (already provided by the places API, but requires changes in browser-interface and unity-renderer)

### Data structure

> To facilitate the addition of the feature when handling favorites we should use the same data structure currently used for Places, with the boolean addition previously mentioned:

Reviewer: Is it really required? Maybe it can be handled by the backend depending on the endpoint call

Author: Currently places displayed in the explore menu are obtained from the lambda for hot scenes `/lambdas/explore/hot-scenes`, this returns a specific data structure with some info about realms and friends connected to that place. 

Because this info is retrieved from the browser-interface (kernel), we could add just the new boolean field `isFavorite` and obtain that from the places API.

Reviewer: I meant just the `isFavorite` field, if places API has specific endpoints for that maybe we can avoid adding it

	public Vector2Int[] userParcels;
}

public string id;
public string name;
public string creator;
public string description;
public string thumbnail;
public Vector2Int baseCoords;
public Vector2Int[] parcels;
public int usersTotalCount;
public bool isPOI;
public Realm[] realms;
public bool isFavorite;
```

## Current limitations

With the current functionalities provided by the Places website it is not possible to favorite all the Decentraland areas due to a restriction that allows only curated Places to be displayed. This together with a legacy code integration of the map in the unity renderer will allow the favorite functionality to be applied only in the places section in the explore menu.

## Decision Journey

In the initial iteration it has been discussed where to store the favorite list, the two options were storing them in the user profile or storing them in Places, the following PROs and CONs made us lean towards the Places option:

Storing favorite in the user profile:
  - PROs
    * Easily readable in unity renderer
    * Simple to obtain favorites of other players
    * Decentralised approach
    * Favorites can be imported in places without user interaction
  - CONs

>     * requires updates on the profile schema

Reviewer: Just a comment here. I think adding a field in the profile schema is not required to make it work (it's an open schema allowing additional properties: https://github.com/decentraland/common-schemas/blob/main/src/platform/profile/profile.ts#L30)

## Glossary

- **Place** A specific parcel or set of parcels identified by a [place UUID](/adr/ADR-186)
- **Places** The structure of database and endpoints used in the places website

## Problem

To enhance the discoverability of the different Decentraland places, there is the need of adding a favorite mechanism. This will allow users to store and retrieve easily a set of their best loved places.
To integrate this, changes at multiple levels are required, from the visual representation in the explorer, to the backend structure that will allow the saving and retrieving of a user favorites.

Following a list of the expected functionalities:

- Users can add or remove a place to their favorite from the explore menu
- Users can navigate to a new tab listing all the places they added to favorites

In a future iteration it will be possible to favorite places from the minimap HUD and also from the map parcel info pop-up, currently not possible due to technical limitations of the renderer map handling and Places API restriction to curated places.
More in the limitations section.

## Solution

### Data storing

> After an initial discussion about whether to store favorites in the user profile or in Places, we are more incline to store everything in Places, more informations about the decision journey at the end.

Reviewer 2: ```suggestion
After an initial discussion about whether to store favorites in the user profile or in Places, we are more inclined to store everything in Places, more information about the decision journey at the end.
```

### Favorite retrieval

> The Places API will provide the required informations through:

Reviewer 2: ```suggestion
The Places API will provide the required information through:
```

- An endpoint to obtain the paginated list of favorites
- An endpoint to add a new favorite for the current user
- An endpoint to remove a favorite for the current user

> In addition a change will be required in the current places data structure in order to show the favorite information in the explore places list with a boolean. (already provided by the places API, but requires changes in browser-interface and unity-renderer)

Reviewer 2: ```suggestion
In addition a change will be required in the current places data structure in order to show the favorite information in the explore places list with a boolean (already provided by the places API, but requires changes in browser-interface and unity-renderer).
```

---
layout: adr
adr: 183
title: Favorite Places
date: 2023-02-07
status: Review
type: Standards Track
spdx-license: CC0-1.0
authors:
  - davidejensen
---

## Abstract

> The document describes the need for a favorite mechanism to enhance the discoverability of different Decentraland places, which requires changes at multiple levels, from the visual representation to the backend structure.

Reviewer 3: ```suggestion
The document describes the need for a "favorite places management" mechanism to enhance the discoverability of different Decentraland scenes and worlds, it requires changes at multiple levels, from the visual representation in the Explorer to the backend services. The state of the document defines the storage place of the favorites as the "Places" server. The decision discards storing the favorites in the user profile.
```

## Glossary

- **Place** A specific parcel or set of parcels identified by a [place UUID](/adr/ADR-186)
- **Places** The structure of database and endpoints used in the places website

## Problem

To enhance the discoverability of the different Decentraland places, there is the need of adding a favorite mechanism. This will allow users to store and retrieve easily a set of their best loved places.
To integrate this, changes at multiple levels are required, from the visual representation in the explorer, to the backend structure that will allow the saving and retrieving of a user favorites.

Following a list of the expected functionalities:

- Users can add or remove a place to their favorite from the explore menu
- Users can navigate to a new tab listing all the places they added to favorites

In a future iteration it will be possible to favorite places from the minimap HUD and also from the map parcel info pop-up, currently not possible due to technical limitations of the renderer map handling and Places API restriction to curated places.
More in the limitations section.

## Solution

### Data storing

> After an initial discussion about whether to store favorites in the user profile or in Places, we are more inclined to store everything in Places, more information about the decision journey at the end.

Reviewer 3: ```suggestion
After an initial discussion about whether to store favorites in the user profile or in the "Places" server, we are more inclined to store everything in the "Places" server, more information about the decision journey at the end.
```

### Favorite retrieval

The Places API will provide the required information through:

- An endpoint to obtain the paginated list of favorites
- An endpoint to add a new favorite for the current user
- An endpoint to remove a favorite for the current user
In addition a change will be required in the current places data structure in order to show the favorite information in the explore places list with a boolean (already provided by the places API, but requires changes in browser-interface and unity-renderer).

### Data structure

To facilitate the addition of the feature when handling favorites we should use the same data structure currently used for Places, with the boolean addition previously mentioned:

> ```

Reviewer 3: ````suggestion
```ts
````
